File: skyspark-scala/src/test/scala/com/akikanellis/skyspark/scala/api/algorithms/bnl/MedianFinderTest.scala
package com.akikanellis.skyspark.scala.api.algorithms.bnl

import com.akikanellis.skyspark.scala.api.algorithms.Point
import com.akikanellis.skyspark.scala.test_utils.{SparkAddOn, UnitSpec}
import org.apache.spark.SparkContext

class MedianFinderTest extends UnitSpec with SparkAddOn {
  private var medianFinder: MedianFinder = _

  before {
    medianFinder = new MedianFinder
  }

  "A single point" should "produce correct median" in withSpark { sc =>
    val points = Seq(Point(1))
    val expectedMedian = Point(0.5)

    assertMedianIsCorrect(sc, points, expectedMedian)
  }

  "1D points" should "produce correct median" in withSpark { sc =>
    val points = Seq(Point(1), Point(3), Point(4), Point(9))
    val expectedMedian = Point(4.5)

    assertMedianIsCorrect(sc, points, expectedMedian)
  }

  "2D points" should "produce correct median" in withSpark { sc =>
    val points = Seq(Point(1, 1), Point(3, 2), Point(4, 7), Point(9, 5))
    val expectedMedian = Point(4.5, 3.5)

    assertMedianIsCorrect(sc, points, expectedMedian)
  }

  "3D points" should "produce correct median" in withSpark { sc =>
    val points = Seq(Point(1, 1, 5), Point(3, 2, 3), Point(4, 7, 1), Point(9, 5, 8))
    val expectedMedian = Point(4.5, 3.5, 4)

    assertMedianIsCorrect(sc, points, expectedMedian)
  }

  private def assertMedianIsCorrect(sc: SparkContext, pointsSeq: Seq[Point], expectedMedian: Point) {
    val points = sc.parallelize(pointsSeq)

    val actualMedian = medianFinder.getMedian(points)

    expectedMedian shouldEqual actualMedian
  }
}

File: skyspark-java/src/test/java/com/akikanellis/skyspark/api/algorithms/bnl/MedianFinderTest.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import com.akikanellis.skyspark.api.test_utils.SparkContextRule;
import org.apache.spark.api.java.JavaRDD;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import static com.akikanellis.skyspark.api.test_utils.assertions.MoreAssertions.assertThat;

public class MedianFinderTest {
    @Rule public SparkContextRule sc = new SparkContextRule();

    private MedianFinder medianFinder;

    @Before public void beforeEach() { medianFinder = new MedianFinder(); }

    @Test public void singlePoint_producesCorrectMedian() {
        JavaRDD<Point> points = sc.parallelize(new Point(1.0));
        Point expectedMedian = new Point(0.5);

        assertExpectedIsMedian(points, expectedMedian);
    }

    @Test public void oneDimensionalPoints_produceCorrectMedian() {
        JavaRDD<Point> points = sc.parallelize(
                new Point(5.2),
                new Point(9.5),
                new Point(4.6)
        );
        Point expectedMedian = new Point(4.75);

        assertExpectedIsMedian(points, expectedMedian);
    }

    @Test public void twoDimensionalPoints_produceCorrectMedian() {
        JavaRDD<Point> points = sc.parallelize(
                new Point(5.2, 2.1),
                new Point(9.5, 1.4),
                new Point(4.6, 4.4)
        );
        Point expectedMedian = new Point(4.75, 2.2);

        assertExpectedIsMedian(points, expectedMedian);
    }

    @Test public void threeDimensionalPoints_produceCorrectMedian() {
        JavaRDD<Point> points = sc.parallelize(
                new Point(5.2, 2.1, 6.7),
                new Point(9.5, 1.4, 7.3),
                new Point(4.6, 4.4, 5.6)
        );
        Point expectedMedian = new Point(4.75, 2.2, 3.65);

        assertExpectedIsMedian(points, expectedMedian);
    }

    private void assertExpectedIsMedian(JavaRDD<Point> points, Point expectedMedian) {
        Point actualMedian = medianFinder.getMedian(points);

        assertThat(actualMedian).isEqualTo(expectedMedian);
    }
}